File: app/src/main/java/com/krystianwsul/checkme/loaders/DomainLoader.java
package com.krystianwsul.checkme.loaders;

import android.content.Context;
import android.support.annotation.NonNull;
import android.support.v4.content.AsyncTaskLoader;

import com.google.firebase.auth.FirebaseAuth;
import com.google.firebase.auth.FirebaseUser;
import com.krystianwsul.checkme.domainmodel.DomainFactory;
import com.krystianwsul.checkme.domainmodel.ObserverHolder;
import com.krystianwsul.checkme.domainmodel.UserInfo;
import com.krystianwsul.checkme.persistencemodel.SaveService;

import junit.framework.Assert;

import java.util.List;

public abstract class DomainLoader<D extends DomainLoader.Data> extends AsyncTaskLoader<D> {
    private D mData;
    private Observer mObserver;

    @NonNull
    private final DomainFactory mDomainFactory;

    @NonNull
    private final FirebaseLevel mFirebaseLevel;

    private final DomainFactory.FirebaseListener mFirebaseListener = domainFactory -> {
        Assert.assertTrue(domainFactory.isConnected());

        if (isStarted())
            forceLoad();
    };

    DomainLoader(@NonNull Context context, @NonNull FirebaseLevel firebaseLevel) {
        super(context);

        mDomainFactory = DomainFactory.getDomainFactory();
        mFirebaseLevel = firebaseLevel;
    }

    @SuppressWarnings("unused")
    abstract String getName();

    @Override
    public final D loadInBackground() {
        if (mFirebaseLevel == FirebaseLevel.NEED && !mDomainFactory.isConnected())
            return null;

        if (mFirebaseLevel == FirebaseLevel.FRIEND && !(mDomainFactory.isConnected() && mDomainFactory.hasFriends()))
            return null;

        return loadDomain(mDomainFactory);
    }

    protected abstract D loadDomain(@NonNull DomainFactory domainFactory);

    // main thread
    @Override
    public void deliverResult(D data) {
        if (data == null)
            return;

        if (isReset())
            return;

        if ((mData == null) || !mData.equals(data)) {
            mData = data;

            if (isStarted())
                super.deliverResult(data);
        }
    }

    // main thread
    @Override
    protected void onStartLoading() {
        if (mData != null)
            deliverResult(mData);

        if (mObserver == null) {
            mObserver = new Observer();
            ObserverHolder.getObserverHolder().addDomainObserver(mObserver);
        }

        if (takeContentChanged() || mData == null) {
            switch (mFirebaseLevel) {
                case NOTHING: {
                    forceLoad();

                    break;
                }
                case WANT: {
                    forceLoad();

                    FirebaseUser firebaseUser = FirebaseAuth.getInstance().getCurrentUser();
                    if (firebaseUser != null && !mDomainFactory.isConnected()) {
                        UserInfo userInfo = new UserInfo(firebaseUser);

                        mDomainFactory.setUserInfo(getContext().getApplicationContext(), SaveService.Source.GUI, userInfo);
                    }

                    break;
                }
                case NEED: {
                    if (mDomainFactory.isConnected()) {
                        forceLoad();
                    } else {
                        FirebaseUser firebaseUser = FirebaseAuth.getInstance().getCurrentUser();
                        if (firebaseUser == null)
                            return;

                        UserInfo userInfo = new UserInfo(firebaseUser);

                        mDomainFactory.setUserInfo(getContext().getApplicationContext(), SaveService.Source.GUI, userInfo);
                        mDomainFactory.addFirebaseListener(mFirebaseListener);
                    }

                    break;
                }
                case FRIEND: {
                    if (mDomainFactory.isConnected() && mDomainFactory.hasFriends()) {
                        forceLoad();
                    } else {
                        FirebaseUser firebaseUser = FirebaseAuth.getInstance().getCurrentUser();
                        if (firebaseUser != null) {
                            UserInfo userInfo = new UserInfo(firebaseUser);

                            mDomainFactory.setUserInfo(getContext().getApplicationContext(), SaveService.Source.GUI, userInfo);
                            mDomainFactory.addFriendFirebaseListener(mFirebaseListener);
                        }
                    }

                    break;
                }
                default:
                    throw new IndexOutOfBoundsException();
            }
        }
    }

    @Override
    protected void onStopLoading() {
        mDomainFactory.removeFirebaseListener(mFirebaseListener);

        cancelLoad();
    }

    @Override
    protected void onReset() {
        onStopLoading();

        if (mData != null)
            mData = null;

        if (mObserver != null) {
            //DomainFactory.getDomainFactory(getContext()).removeDomainObserver(mObserver);
            mObserver = null;
        }
    }

    public class Observer {
        public void onDomainChanged(@NonNull List<Integer> dataIds) {
            if (mData != null && dataIds.contains(mData.DataId))
                return;

            onContentChanged();
        }
    }

    public abstract static class Data {
        private static int sDataId = 1;

        public final int DataId;

        private static int getNextId() {
            return sDataId++;
        }

        public Data() {
            DataId = getNextId();
        }
    }

    enum FirebaseLevel {
        NOTHING,
        WANT,
        NEED,
        FRIEND
    }
}

File: app/src/main/java/com/krystianwsul/checkme/loaders/DomainLoader.kt
package com.krystianwsul.checkme.loaders

import android.content.Context
import android.support.v4.content.AsyncTaskLoader
import com.google.firebase.auth.FirebaseAuth
import com.krystianwsul.checkme.domainmodel.DomainFactory
import com.krystianwsul.checkme.domainmodel.ObserverHolder
import com.krystianwsul.checkme.domainmodel.UserInfo
import com.krystianwsul.checkme.firebase.RemoteFriendFactory
import com.krystianwsul.checkme.persistencemodel.SaveService
import com.krystianwsul.checkme.viewmodels.DomainObserver

abstract class DomainLoader<D : DomainData>(context: Context, private val firebaseLevel: FirebaseLevel) : AsyncTaskLoader<D>(context) {

    private var data: D? = null

    private var observer: Observer? = null

    private val domainFactory = DomainFactory.getDomainFactory()

    private val firebaseListener = { domainFactory: DomainFactory ->
        check(domainFactory.isConnected)

        if (isStarted)
            forceLoad()
    }

    abstract val name: String

    override fun loadInBackground(): D? {
        if (firebaseLevel == FirebaseLevel.NEED && !domainFactory.isConnected)
            return null

        return if (firebaseLevel == FirebaseLevel.FRIEND && !(domainFactory.isConnected && RemoteFriendFactory.hasFriends())) null else loadDomain(domainFactory)
    }

    protected abstract fun loadDomain(domainFactory: DomainFactory): D

    // main thread
    override fun deliverResult(data: D?) {
        if (data == null)
            return

        if (isReset)
            return

        if (this.data == null || this.data != data) {
            this.data = data

            if (isStarted)
                super.deliverResult(data)
        }
    }

    // main thread
    override fun onStartLoading() {
        if (data != null)
            deliverResult(data)

        if (observer == null) {
            observer = Observer()
            ObserverHolder.addDomainObserver(observer!!)
        }

        if (takeContentChanged() || data == null) {
            when (firebaseLevel) {
                FirebaseLevel.NOTHING -> forceLoad()
                FirebaseLevel.WANT -> {
                    forceLoad()

                    val firebaseUser = FirebaseAuth.getInstance().currentUser
                    if (firebaseUser != null && !domainFactory.isConnected) {
                        val userInfo = UserInfo(firebaseUser)

                        domainFactory.setUserInfo(context.applicationContext, SaveService.Source.GUI, userInfo)
                    }
                }
                FirebaseLevel.NEED -> {
                    if (domainFactory.isConnected) {
                        forceLoad()
                    } else {
                        val firebaseUser = FirebaseAuth.getInstance().currentUser ?: return

                        val userInfo = UserInfo(firebaseUser)

                        domainFactory.setUserInfo(context.applicationContext, SaveService.Source.GUI, userInfo)
                        domainFactory.addFirebaseListener(firebaseListener)
                    }
                }
                FirebaseLevel.FRIEND -> {
                    if (domainFactory.isConnected && RemoteFriendFactory.hasFriends()) {
                        forceLoad()
                    } else {
                        FirebaseAuth.getInstance().currentUser?.let {
                            val userInfo = UserInfo(it)

                            domainFactory.setUserInfo(context.applicationContext, SaveService.Source.GUI, userInfo)
                            RemoteFriendFactory.addFriendListener { firebaseListener(DomainFactory.getDomainFactory()) }
                        }
                    }
                }
            }
        }
    }

    override fun onStopLoading() {
        domainFactory.removeFirebaseListener(firebaseListener)

        cancelLoad()
    }

    override fun onReset() {
        onStopLoading()

        data = null
        observer = null
    }

    inner class Observer : DomainObserver {

        override fun onDomainChanged(dataIds: List<Int>) {
            if (data?.let { dataIds.contains(it.dataId) } == true)
                return

            onContentChanged()
        }
    }
}
